src/mailmap.c

```diff
@@ -0,0 +1,335 @@
+			entry->real_email = entry_data;
+			entry_data += real_email.size + 1;
+		}
+		if (replace_name.size > 0) {
+			memcpy(entry_data, replace_name.ptr, replace_name.size);
+			entry->replace_name = entry_data;
+			entry_data += replace_name.size + 1;
+		}
+		/* replace_email is always non-null */
+		memcpy(entry_data, replace_email.ptr, replace_email.size);
+		entry->replace_email = entry_data;
+
+		error = git_vector_insert(&mm->entries, entry);
+		if (error < 0)
+			goto cleanup;
+		entry = NULL;
+	}
+
+	/* fill in *out, and make sure we don't free our mailmap */
+	*out = mm;
+	mm = NULL;
+
+cleanup:
+	git__free(entry);
+	git_mailmap_free(mm);
+
+	/* We never allocate data in these buffers, but better safe than sorry */
+	git_buf_free(&real_name);
+	git_buf_free(&real_email);
+	git_buf_free(&replace_name);
+	git_buf_free(&replace_email);
+	return error;
+}
+
+void git_mailmap_free(git_mailmap *mailmap)
+{
+	if (!mailmap)
+		return;
+
+	git_vector_free_deep(&mailmap->entries);
+	git__free(mailmap);
+}
+
+void git_mailmap_resolve(
+	const char **name_out, const char **email_out,
+	const git_mailmap *mailmap,
+	const char *name, const char *email)
+{
+	const git_mailmap_entry *entry = NULL;
+	assert(name && email);
+
+	*name_out = name;
+	*email_out = email;
+
+	if (!mailmap)
+		return;
+
+	entry = git_mailmap_entry_lookup(mailmap, name, email);
+	if (entry) {
+		if (entry->real_name)
+			*name_out = entry->real_name;
+		if (entry->real_email)
+			*email_out = entry->real_email;
+	}
+}
+
+const git_mailmap_entry *git_mailmap_entry_lookup(
+	const git_mailmap *mailmap, const char *name, const char *email)
+{
+	size_t i;
+	git_mailmap_entry *entry;
+	assert(name && email);
+
+	if (!mailmap)
+		return NULL;
+
+	git_vector_foreach(&mailmap->entries, i, entry) {
+		if (!git__strcmp(email, entry->replace_email) &&
+		    (!entry->replace_name || !git__strcmp(name, entry->replace_name))) {
+			return entry;
+		}
+	}
+
+	return NULL;
+}
+
+const git_mailmap_entry *git_mailmap_entry_byindex(
+	const git_mailmap *mailmap, size_t idx)
+{
+	if (mailmap)
+		return git_vector_get(&mailmap->entries, idx);
+	return NULL;
+}
+
+size_t git_mailmap_entry_count(const git_mailmap *mailmap)
+{
+	if (mailmap)
+		return git_vector_length(&mailmap->entries);
+	return 0;
+}
+
+static int mailmap_from_bare_repo(git_mailmap **mailmap, git_repository *repo)
+{
+	git_reference *head = NULL;
+	git_object *tree = NULL;
+	git_blob *blob = NULL;
+	git_buf content = GIT_BUF_INIT;
+	int error;
+
+	assert(git_repository_is_bare(repo));
+
+	/* In bare repositories, fall back to reading from HEAD's tree */
+	error = git_repository_head(&head, repo);
+	if (error < 0)
+		goto cleanup;
+
+	error = git_reference_peel(&tree, head, GIT_OBJ_TREE);
+	if (error < 0)
+		goto cleanup;
+
+	error = git_object_lookup_bypath(
+		(git_object **) &blob, tree, MAILMAP_FILE, GIT_OBJ_BLOB);
+	if (error < 0)
+		goto cleanup;
+
+	error = git_blob_filtered_content(&content, blob, MAILMAP_FILE, false);
```

Comment:
I don't really think it necessary to have that blob filtered. Just using `git_blob__getbuf(&content, blob)` should be fine. Or are there people who define a custom filter that writes the mailmap dynamically... :scream: Users can get creative, but until somebody complains I'd just go for the unfiltered content.

Reply:
This raises a question: can we properly parse mailmap files with Windows-style line endings?  If not, then we should absolutely _not_ be using the filtered content.

I think that the goal is to use the mailmap as it would exist in the working directory, which makes sense.  But can git cope with mailmaps with Windows-style line endings?  Does it use the filtered content or the content directly from the repository?

Reply:
```
$ echo -e 'test <foo@bar.com>\r\nother <other@bar.com>\r\n' | git check-mailmap --stdin
test <foo@bar.com>
other <other@bar.com>
fatal: unable to parse contact:
```

So yes, it does cope with Windows-style line endings. The last line is simply that it cannot parse an empty line.

Reply:
That being said, I think _we_ as in the libgit2 project cannot currently parse them with Windows-style endings.

Reply:
Let's make sure that we can, please.  👍